v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/jdbc/iowrapper/JdbcIoWrapper.java

```diff
@@ -0,0 +1,182 @@
+
+public final class JdbcIoWrapper implements IoWrapper {
+  private final ImmutableMap<SourceTableReference, PTransform<PBegin, PCollection<SourceRow>>>
+      tableReaders;
+  private final SourceSchema sourceSchema;
+
+  public static JdbcIoWrapper of(JdbcIOWrapperConfig config) {
+    var dataSourceConfiguration = getDataSourceConfiguration(config);
+
+    var sourceSchema = getSourceSchema(config, dataSourceConfiguration);
+    var tableReaders = buildTableReaders(config, dataSourceConfiguration, sourceSchema);
+    return new JdbcIoWrapper(tableReaders, sourceSchema);
+  }
+
+  @Override
+  public ImmutableMap<SourceTableReference, PTransform<PBegin, PCollection<SourceRow>>>
+      getTableReaders() {
+    return this.tableReaders;
+  }
+
+  @Override
+  public SourceSchema discoverTableSchema() {
+    return this.sourceSchema;
+  }
+
+  static ImmutableMap<SourceTableReference, PTransform<PBegin, PCollection<SourceRow>>>
+      buildTableReaders(
+          JdbcIOWrapperConfig config,
+          DataSourceConfiguration dataSourceConfiguration,
+          SourceSchema sourceSchema) {
+    return config.tableConfigs().stream()
+        .map(
+            tableConfig -> {
+              SourceTableSchema sourceTableSchema =
+                  findSourceTableSchema(sourceSchema, tableConfig);
+              return Map.entry(
+                  SourceTableReference.builder()
+                      .setSourceSchemaReference(sourceSchema.schemaReference())
+                      .setSourceTableName(sourceTableSchema.tableName())
+                      .setSourceTableSchemaUUID(sourceTableSchema.tableSchemaUUID())
+                      .build(),
+                  getJdbcIO(config, dataSourceConfiguration, tableConfig, sourceTableSchema));
+            })
+        .collect(ImmutableMap.toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+  }
+
+  static SourceTableSchema findSourceTableSchema(
+      SourceSchema sourceSchema, TableConfig tableConfig) {
+    return sourceSchema.tableSchemas().stream()
+        .filter(schema -> schema.tableName().equals(tableConfig.tableName()))
+        .findFirst()
+        .orElseThrow();
+  }
+
+  static SourceSchema getSourceSchema(
+      JdbcIOWrapperConfig config, DataSourceConfiguration dataSourceConfiguration) {
+    SchemaDiscovery schemaDiscovery =
+        new SchemaDiscoveryImpl(config.dialectAdapter(), config.schemaDiscoveryBackOff());
+    var sourceSchemaBuilder =
+        SourceSchema.builder().setSchemaReference(config.sourceSchemaReference());
+    DataSource dataSource = dataSourceConfiguration.buildDatasource();
+    ImmutableList<String> tables =
+        config.tableConfigs().stream()
+            .map(TableConfig::tableName)
+            .collect(ImmutableList.toImmutableList());
+    var tableSchemas =
+        schemaDiscovery.discoverTableSchema(dataSource, config.sourceSchemaReference(), tables);
+    tableSchemas.entrySet().stream()
```

Comment:
nit: code maybe a lot shorter and more readable by just using a simple for each instead of lambdas here.

Reply:
Thanks. Makes sense. This ended up being lot more nested than I had throught earlier :(

I will cover it as a part of followup tracked in b/339174207